Developer Workflow UX Document

technosophos opened the pull request:
This draft document outlines the user experience for Helm users. The document is a combination of three earlier documents (list of operations, local developer workflow, team development workflow). The UML sequence diagram source code (PlantUML markup) is also included in the commit.

technosophos committed:
Add workflow document.

This addresses #236.

technosophos committed:
Merged in CLI UX document.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+- The server returns a stub file
+- The client writes the file to disk
+
+#### Deploy with params:
```
Nice. Can we also deploy with parameters on the command line, as in current dm client?

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+If we allow the user to pass in a name (overriding the generated name),
+then the server must first guarantee that this name is unique.
+
+#### Get the info about named deployment.
```
Would be useful to return the name of the chart, also.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+- The server coalesces the values into the existing instance and then
+  restarts.
+
+#### Uninstall:
```
Uninstall seems odd, since we didn't install taco-tuesday. Delete might be better.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+- The server returns the values file used to generate the instance.
+- The client writes this to a file (or, perhaps, to STDOUT)
+
+#### Get fully generated manifest files
```
Currently, we maintain a history of manifests. We should probably provide a flag to let the user download only the latest, which represents the current deployment, or download the full history. Not sure which should be the default. Maybe just the latest, with a flag to get the history.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+- The client sends the DELETE instance name command
+- The API server destroys the resource
+
+### Power User Features
```
Before we get to power user features, there are some additional use cases for the tire kicker:
* List all my deployed charts in this cluster.
* List all the chars available in a given helm repository.
* List all the known helm repositories.
* Add a helm repository to the known list.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+- The server assigns a random name `wonky-panda`, fetches the chart from
+  object storage, and goes about the deployment process.
+
+#### Find out about params:
```
Let this command and the following commands take a list of charts.

runseb commented on docs/workflow/developer-workflows.md:
````diff
@@ -0,0 +1,441 @@
+
+#### Get values for an app:
+```
+$ helm get-values taco-tuesday
````
Would be nice to have a sentence based CLI, akin to kubectl. If we define all verbs and objects ...we can have a very intuitive CLI.

helm get charts
helm get repos
helm get manifests

helm delete charts
..etc

jackgr commented on docs/workflow/developer-workflows.md:
````diff
@@ -0,0 +1,441 @@
+values.
+
+```
+$ helm create mychart
````
Would be nice to combine this use case with retrieving the values for one or more deployed charts, or perhaps generating a chart from a config file, which the user downloaded with get values.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+workflow, but do have occasional uses. Because they are less common,
+many of these are grouped into sections of subcommands.
+
+- `helm dm install|uninstall|status|target`: Get information about a
```
Could use more description here, show how these commands fit into workflows.

jackgr commented on docs/workflow/developer-workflows.md:
```diff
@@ -0,0 +1,441 @@
+- `helm release`: Push a package to the repository. Client sends directly
+  to repository.
+- `helm list`: List deployments. Processed on the manager
+- `helm search`: Search for available charts matching a given pattern.
```
These two commands are likely to be part of the daily workflow.

technosophos committed:
Added details on numerous other commands.

adamreese commented on docs/workflow/developer-workflows.md:
````diff
@@ -0,0 +1,648 @@
+#### Adding credentials
+
+```
+$ helm credential add TOKEN SECRET
````
Are credentials repository specific?

technosophos committed:
Added descriptions to a few of the additional commands.

jackgr commented:
@technosophos there are some significant comments not yet addressed in this PR. Given the importance of this document, it would be great to get them addressed ASAP, so that we can LGTM and merge.

technosophos committed:
Changes to commands based on "helm NOUN VERB" form

This normalizes several commands to the "helm NOUN VERB" form we
have been talking about.

technosophos commented:
To avoid rebase hell, closing this and opening a duplicate.
